### PR TITLE
[stable/mysql] Update mysql user passwords on secrets change

### PR DESCRIPTION
**Pull request by komljen**
If you don't specify MySQL root and other users passwords they are auto-generated. However, if you run the Helm upgrade the new auto-generated secrets will become invalid. With this fix, it will be possible to change user-specified password also on Helm upgrade.

### CODE REVIEW

**Comment by rjkernick** [comment]
Please resolve the merge conflicts

**Comment by Miouge1** [comment]
Thanks @komljen I was looking for this exact feature.

**Comment by unguiculus** [comment]
@viglesiasce Do we really want to do this? We have this problem with all charts that auto-generate passwords. IMHO you should always set a password explicitly for purposes other then simple testing. Alternatively, there's the `--reuse-values` flag for `helm upgrade`. Automatically changing passwords on upgrade is not something I would expect.

/hold

**Comment by komljen** [comment]
> IMHO you should always set a password explicitly for purposes other then simple testing. 

I agree on this one, but when I tried this chart for the first time I got confused why my apps started to fail after I did the upgrade of umbrella chart where mysql was a dependency. I think that Jenkins chart is changing the password on upgrade if not set also.

**Comment by Miouge1** [comment]
@unguiculus is `--reuse-values` really an alternative for this issue?

I have tested it like this:
```
helm upgrade --install mysql stable/mysql --set imageTag=5.7.10 --set persistence.enabled=false --reuse-values
kubectl exec $(kubectl get pods -l app=mysql-mysql -o jsonpath='{.items[0].metadata.name}') env | grep PASSWORD

helm upgrade --install mysql stable/mysql --set imageTag=5.7.11 --set persistence.enabled=false --reuse-values
kubectl exec $(kubectl get pods -l app=mysql-mysql -o jsonpath='{.items[0].metadata.name}') env | grep PASSWORD
```

which shows that passwords are re-randomized between both deploys. It would make sense since when the password is not set, the randAlphNum doesn't save its output to values no?

**Comment by viglesiasce** [comment]
Another approach would be to do the secret creation only on first install using the release number as a gate for example:

```
{{ if eq .Release.Revision 1 }}
# Secret YAML
{{ end }}
```

**Comment by komljen** [comment]
Good point, but if we specify our own password and then we want to change it, we could just use `--set` with helm upgrade and our MySQL would have a new password. Currently, we can not change the MySQL password with helm upgrade and this pull request solves that.

**Comment by scottrigby** [comment]
>We have this problem with all charts that auto-generate passwords.

This has been on my mind for various charts - as well as those charts which depend on the ones auto-generating passwords. It would be good to agree on an approach across charts that have this problem. Can we address this at our next charts meeting? I have created #5167 to track this, as well as references a number of related issues/PRs/docs there.

File: stable/mysql/Chart.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 name: mysql
-version: 0.3.7
+version: 0.3.8
```
Comment (allanlewis) [inline stable/mysql/Chart.yaml:2]: Perhaps this justifies a "minor" rather than a "patch" version bump?

**Comment by cliedeman** [comment]
Created existingSecret option as an alternative solution - #5783

**Comment by komljen** [comment]
This pull request solves the problem of updating the password with helm upgrade command. It doesn't matter if the password is auto-generated or set by the user.

File: stable/mysql/Chart.yaml (hunk 1 above)
Comment (komljen) [reply stable/mysql/Chart.yaml:2]: Yeah, this should be minor.

**Comment by benlangfeld** [comment]
What prevented this from being merged?

**Comment by Tarjei400** [comment]
Bumping up @benlangfeld question, that would be really nice to have feature in a chart

**Comment by Tarjei400** [comment]
Just in case someone has same issue, it is possible to make exact this thing using extraContainers, extraVolumeMounts etc. below example how to parametrize a chart: 
`args: [ "--init-file=/tmp/files/mysql-init" ]
#Update database credentials with newly generated ones in secret
#1. Create initialization script
#2. Write env variables to that initization script and write it to shared volume
#3. Adds extra argument to get init file from shared volume!

initializationFiles:
  mysql-init: |-
    UPDATE mysql.user SET authentication_string = PASSWORD('MYSQL_ROOT_PASSWORD')
    WHERE User = 'root' AND Host = '%';
    UPDATE mysql.user SET authentication_string = PASSWORD('MYSQL_PASSWORD')
    WHERE User = 'MYSQL_USER' AND Host = '%';
    FLUSH PRIVILEGES;

extraInitContainers: |
  - name: "update-mysql-init"
    image: "busybox:1.25.0"
    imagePullPolicy: {{ .Values.imagePullPolicy | quote }}
    command: [
      "sh",
      "-ce",
      "cp /docker-entrypoint-initdb.d/mysql-init /tmp/files/mysql-init &&
         sed -i \"s/MYSQL_ROOT_PASSWORD/$MYSQL_ROOT_PASSWORD/g\" /tmp/files/mysql-init &&
         sed -i \"s/MYSQL_PASSWORD/$MYSQL_PASSWORD/g\" /tmp/files/mysql-init &&
         sed -i \"s/MYSQL_USER/$MYSQL_USER/g\" /tmp/files/mysql-init" ]
    env:
      - name: MYSQL_ROOT_PASSWORD
        valueFrom:
          secretKeyRef:
            name: {{ template "mysql.fullname" . }}
            key: mysql-root-password
      - name: MYSQL_PASSWORD
        valueFrom:
          secretKeyRef:
            name: {{ template "mysql.fullname" . }}
            key: mysql-password
    volumeMounts:
      - name: migrations
        mountPath: /docker-entrypoint-initdb.d
      - name: tmp-files
        mountPath: /tmp/files

extraVolumes: |
  - name: tmp-files
    emptyDir: {}

extraVolumeMounts: |
  - name: tmp-files
    mountPath: /tmp/files`

**Comment by kylezs** [comment]
@Tarjei400 's solution did not work for me, perhaps because I was using a persistent volume. 
Instead, I:
- Noted the current secrets being used for the mysql root and user (if you have a different one)
- Updated the secrets
- Restarted the pods (so the env vars updated)
- port-forwarded the mysql pod
- connected a local mysql client to the pod (using previously noted credentials of root user)
- used the mysql client to update the passwords 
`ALTER USER 'userName'@'localhost' IDENTIFIED BY 'New-Password-Here';`

**Pull request closed without merging**